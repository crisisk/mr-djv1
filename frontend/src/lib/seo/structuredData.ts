export interface StructuredDataScript {
  id: string;
  json: string;
}

const BASE_URL = "https://staging.sevensa.nl/";
const LOGO_URL = `${BASE_URL}assets/images/logo.png`;
const BUSINESS_ID = `${BASE_URL}#mister-dj`;
const ORGANIZATION_ID = `${BASE_URL}#organization`;
const SERVICE_ID = `${BASE_URL}#wedding-and-event-dj-service`;

const areaServed = ["Brabant", "Eindhoven", "Tilburg", "Den Bosch", "Zuid-Nederland"];

const sameAs = [
  "https://www.facebook.com/misterdj",
  "https://www.instagram.com/misterdj",
  "https://www.linkedin.com/company/mister-dj",
];

const localBusinessSchema = {
  "@context": "https://schema.org",
  "@type": "LocalBusiness",
  "@id": BUSINESS_ID,
  name: "Mister DJ",
  description:
    "Mister DJ levert dj's en live sax combinaties met 100% dansgarantie voor bruiloften, bedrijfsfeesten en private events.",
  image: LOGO_URL,
  url: BASE_URL,
  telephone: "+31-40-8422594",
  priceRange: "€€",
  address: {
    "@type": "PostalAddress",
    streetAddress: "Kapteijnlaan 17",
    postalCode: "5505 AV",
    addressLocality: "Veldhoven",
    addressRegion: "NB",
    addressCountry: "NL",
  },
  areaServed,
  sameAs,
};

const organizationSchema = {
  "@context": "https://schema.org",
  "@type": "Organization",
  "@id": ORGANIZATION_ID,
  name: "Mister DJ",
  url: BASE_URL,
  logo: LOGO_URL,
  sameAs,
  contactPoint: [
    {
      "@type": "ContactPoint",
      telephone: "+31-40-8422594",
      contactType: "customer service",
      areaServed: "NL",
      availableLanguage: ["nl", "en"],
    },
  ],
};

const serviceSchema = {
  "@context": "https://schema.org",
  "@type": "Service",
  "@id": SERVICE_ID,
  name: "Mister DJ Bruiloft & Event DJ",
  serviceType: "Wedding and corporate event DJ services",
  description:
    "Volledige DJ-service met maatwerk licht- en geluidsoplossingen voor bruiloften, bedrijfsfeesten en exclusieve events in Zuid-Nederland.",
  provider: {
    "@type": "Organization",
    "@id": ORGANIZATION_ID,
    name: "Mister DJ",
  },
  areaServed: [
    {
      "@type": "AdministrativeArea",
      name: "Brabant",
    },
    {
      "@type": "City",
      name: "Eindhoven",
    },
    {
      "@type": "City",
      name: "Tilburg",
    },
    {
      "@type": "City",
      name: "Den Bosch",
    },
    {
      "@type": "State",
      name: "Zuid-Nederland",
    },
  ],
  offers: {
    "@type": "Offer",
    priceCurrency: "EUR",
    availability: "https://schema.org/InStock",
    url: BASE_URL,
    eligibleRegion: {
      "@type": "Country",
      name: "NL",
    },
  },
};

export const structuredDataSchemas = {
  localBusiness: localBusinessSchema,
  organization: organizationSchema,
  service: serviceSchema,
} as const;

export type StructuredDataKey = keyof typeof structuredDataSchemas;

export const toJsonLd = (schema: object): string => JSON.stringify(schema, null, 2);

export const getStructuredDataScripts = (): StructuredDataScript[] =>
  (Object.keys(structuredDataSchemas) as StructuredDataKey[]).map((key) => ({
    id: `schema-${key}`,
    json: toJsonLd(structuredDataSchemas[key]),
  }));

export const injectStructuredData = (doc?: Document): void => {
  const targetDocument = doc ?? (typeof document !== "undefined" ? document : undefined);
  if (!targetDocument) {
    return;
  }

  const head = targetDocument.head ?? targetDocument.getElementsByTagName("head")[0];
  if (!head) {
    return;
  }

  getStructuredDataScripts().forEach(({ id, json }) => {
    let script = targetDocument.getElementById(id) as HTMLScriptElement | null;
    if (!script) {
      script = targetDocument.createElement("script");
      script.type = "application/ld+json";
      script.id = id;
      head.appendChild(script);
    }

    script.textContent = json;
  });
};
