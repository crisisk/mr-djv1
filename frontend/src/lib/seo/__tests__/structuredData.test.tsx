import { afterEach, describe, expect, it } from 'vitest'
import { injectStructuredData, getStructuredDataScripts, structuredDataSchemas } from '../structuredData'

afterEach(() => {
  document.head.innerHTML = ''
})

describe('structuredData', () => {
  it('returns JSON-LD scripts for all schemas', () => {
    const scripts = getStructuredDataScripts()

    expect(scripts).toHaveLength(Object.keys(structuredDataSchemas).length)
    scripts.forEach(({ json }) => {
      expect(() => JSON.parse(json)).not.toThrow()
    })
  })

  it('injects structured data scripts into the head exactly once', () => {
    injectStructuredData(document)
    injectStructuredData(document)

    const scripts = document.querySelectorAll("script[type='application/ld+json'][id^='schema-']")

    expect(scripts).toHaveLength(3)
    expect(document.getElementById('schema-organization')).toBeInTheDocument()
    expect(document.getElementById('schema-service')).toBeInTheDocument()
    expect(document.getElementById('schema-localBusiness')).toBeInTheDocument()
    expect(document.head).toMatchSnapshot()

    scripts.forEach((script) => {
      expect(() => JSON.parse(script.textContent ?? '')).not.toThrow()
    })
  })
})
