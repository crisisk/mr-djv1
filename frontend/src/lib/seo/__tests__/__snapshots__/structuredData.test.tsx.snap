// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`structuredData > injects structured data scripts into the head exactly once 1`] = `
<head>
  <script
    id="schema-localBusiness"
    type="application/ld+json"
  >
    {
  "@context": "https://schema.org",
  "@type": "LocalBusiness",
  "@id": "https://staging.sevensa.nl/#mister-dj",
  "name": "Mister DJ",
  "description": "Mister DJ levert dj's en live sax combinaties met 100% dansgarantie voor bruiloften, bedrijfsfeesten en private events.",
  "image": "https://staging.sevensa.nl/assets/images/logo.png",
  "url": "https://staging.sevensa.nl/",
  "telephone": "+31-40-8422594",
  "priceRange": "€€",
  "address": {
    "@type": "PostalAddress",
    "streetAddress": "Kapteijnlaan 17",
    "postalCode": "5505 AV",
    "addressLocality": "Veldhoven",
    "addressRegion": "NB",
    "addressCountry": "NL"
  },
  "areaServed": [
    "Brabant",
    "Eindhoven",
    "Tilburg",
    "Den Bosch",
    "Zuid-Nederland"
  ],
  "sameAs": [
    "https://www.facebook.com/misterdj",
    "https://www.instagram.com/misterdj",
    "https://www.linkedin.com/company/mister-dj"
  ]
}
  </script>
  <script
    id="schema-organization"
    type="application/ld+json"
  >
    {
  "@context": "https://schema.org",
  "@type": "Organization",
  "@id": "https://staging.sevensa.nl/#organization",
  "name": "Mister DJ",
  "url": "https://staging.sevensa.nl/",
  "logo": "https://staging.sevensa.nl/assets/images/logo.png",
  "sameAs": [
    "https://www.facebook.com/misterdj",
    "https://www.instagram.com/misterdj",
    "https://www.linkedin.com/company/mister-dj"
  ],
  "contactPoint": [
    {
      "@type": "ContactPoint",
      "telephone": "+31-40-8422594",
      "contactType": "customer service",
      "areaServed": "NL",
      "availableLanguage": [
        "nl",
        "en"
      ]
    }
  ]
}
  </script>
  <script
    id="schema-service"
    type="application/ld+json"
  >
    {
  "@context": "https://schema.org",
  "@type": "Service",
  "@id": "https://staging.sevensa.nl/#wedding-and-event-dj-service",
  "name": "Mister DJ Bruiloft & Event DJ",
  "serviceType": "Wedding and corporate event DJ services",
  "description": "Volledige DJ-service met maatwerk licht- en geluidsoplossingen voor bruiloften, bedrijfsfeesten en exclusieve events in Zuid-Nederland.",
  "provider": {
    "@type": "Organization",
    "@id": "https://staging.sevensa.nl/#organization",
    "name": "Mister DJ"
  },
  "areaServed": [
    {
      "@type": "AdministrativeArea",
      "name": "Brabant"
    },
    {
      "@type": "City",
      "name": "Eindhoven"
    },
    {
      "@type": "City",
      "name": "Tilburg"
    },
    {
      "@type": "City",
      "name": "Den Bosch"
    },
    {
      "@type": "State",
      "name": "Zuid-Nederland"
    }
  ],
  "offers": {
    "@type": "Offer",
    "priceCurrency": "EUR",
    "availability": "https://schema.org/InStock",
    "url": "https://staging.sevensa.nl/",
    "eligibleRegion": {
      "@type": "Country",
      "name": "NL"
    }
  }
}
  </script>
</head>
`;
